scripting to run enki on ABL

kswanson33 opened the pull request:
Enables user to generate a book list from ABL and (optionally) a user-maintained list of non-ABL books.

openstax/cnx-recipes#4969

kswanson33 commented:
Pros of using ruby: 
- it's easy & tidy
- bash integration

Cons:
- anyone who runs `update_books.rb` needs to have ruby installed

To mitigate this, I could do something with a mini Docker environment, or something else. Thoughts?

kswanson33 commented:
The directory name for this collection isn't set in stone. Current name: bap (Big A Pipeline)

other options:
- cross_library
- automate_for_lib
- enki_on_booklist
- enki_on_lib

TylerZeroMaster commented on bap/update_books.rb:
```diff
@@ -0,0 +1,29 @@
+abl_hash['approved_books'].each do |book|
+  repo = book['repository_name']
+  book['versions'][0]['commit_metadata']['books'].each do |volume| 
+    File.write(out_file, "#{repo} #{volume['slug']}\n", mode: 'a')
```
You could do this with jq instead of ruby if you're interested:
```bash
abl="$(curl -Ss 'https://raw.githubusercontent.com/openstax/content-manager-approved-books/main/approved-book-list.json')"
echo "$abl" | jq -r '.approved_books|.[]|.repository_name as $repo|.versions[0].commit_metadata.books|.[]|$repo + " " + .slug'
```
The output is something like

> osbooks-us-history us-history
> osbooks-microbiology microbiology
> osbooks-university-physics-bundle university-physics-volume-1
> osbooks-university-physics-bundle university-physics-volume-2
> osbooks-university-physics-bundle university-physics-volume-3
> osbooks-prealgebra-bundle prealgebra
> osbooks-prealgebra-bundle elementary-algebra
> osbooks-prealgebra-bundle intermediate-algebra

philschatz replied:
Just formatting the 2nd line a bit:

```sh
abl="$(curl -Ss 'https://raw.githubusercontent.com/openstax/content-manager-approved-books/main/approved-book-list.json')"

query=$(cat << 'EOF'
    .approved_books                         # Select approved_books
    | .[]                                   # Foreach (approved_book) in approved_books
    | .repository_name as $repo             #   Assign $repo to be the value of .repository_name
    | .versions[0].commit_metadata.books    #   Select books in version[0].commit_metadata
    | .[]                                   #   Foreach (book) in books
    | $repo + " " + .slug                   #     Print the $repo and the .slug
EOF
)

echo "$abl" | jq -r "$query"
```

kswanson33 commented:
- added an `update_books.sh` as per tyler's suggestion. 
- `update_books.rb` still exists, and now removes duplicate books. any suggestions on how to eliminate duplicates in bash?
- updated `run_enki_on_all` to write logs from enki runs to files, & display status of each run. example:
<img width="432" alt="image" src="https://user-images.githubusercontent.com/26280712/220376446-4ee4e2d9-29c4-4c41-a631-0ab679d7597a.png">

TylerZeroMaster commented on cross-lib/run_enki_on_all.sh:
```diff
@@ -0,0 +1,51 @@
+  ./enki --data-dir ./data/$3-$1 --command $1 --repo openstax/$2 --book-slug $3 --style default --ref main &> cross-lib/logs/$2-$3.txt
+  exit=$?
+  stop_time=$(date +%s)
+  elapsed_formatted="$(date -u -r $(($stop_time - $start_time)) +%T)"
```
The date command here did not work for me at all. The -r option pulls the last modified date from a file. This one did work. This creates a date that is `elapsed_time` after epoch and then pulls out the hours, minutes, and seconds.
```bash
elapsed_formatted="$(date --date="@$((stop_time - start_time))" +%H:%M:%S)"
```

kswanson33 replied:
hmm, this suggestion isn't working for me 😅 i believe it's a matter of [linux](https://man7.org/linux/man-pages/man1/date.1.html) vs [mac](https://ss64.com/osx/date.html) `date`.

have you encountered this sort of issue before? any thoughts on what to do?

kswanson33 replied:
update: tried [this](https://github.com/openstax/enki/pull/200/files#diff-1aba1be9b3f156d66a15a2ccf9725ff074ff6a413320298f3762a3d811884ddaR33), does that work for you?

TylerZeroMaster replied:
Sorry for the late reply, I think you need to @ sign to indicate that the time is in second.
```bash
$ date --date="1" +%H:%M:%S
01:00:00
$ date --date="@1" +%H:%M:%S
00:00:01
```

TylerZeroMaster commented on cross-lib/run_enki_on_all.sh:
```diff
@@ -0,0 +1,51 @@
+  echo "" | run_and_log_enki $arg_command $repo $slug || true
+done <$all_books
+
+# final time report (& failures?)
```
When I wanted to stop building books, it kept going 😧. This is one way to fix that problem.
```bash
# Do not keep looping if I interrupt
trap 'exit 0' INT
```

TylerZeroMaster commented on cross-lib/run_enki_on_all.sh:
```diff
@@ -0,0 +1,51 @@
+  repo=${line%%' '*}
+  slug=${line##*' '}
+  # https://stackoverflow.com/a/35208546
+  echo "" | run_and_log_enki $arg_command $repo $slug || true
```
```bash
echo "" | ( run_and_log_enki $arg_command $repo $slug || true ) 2>&1 | tee "cross-lib/logs/$repo-$slug.txt"
```

I like having the exit code and the elapsed time in the log files.

kswanson33 replied:
i agree it's nice to have the logging in the terminal too. on the other hand, someone looping over a lot of books might want to know what book they're on in the list and how the list is going. the logging about each book's build obscures the information about the status of the list. maybe we can discuss preferred logging behavior more in depth?

TylerZeroMaster commented on cross-lib/run_enki_on_all.sh:
```diff
@@ -0,0 +1,51 @@
+[[ $arg_command ]] || ( echo "ERROR: A command was not provided. Typical examples are 'all-git-pdf' or 'all-git-web' or 'all-git-epub'" && exit 1 )
+
+all_books="cross-lib/book_data/AUTO_books.txt"
+test -f $all_books || ( echo "ERROR: Book list not found at ${all_books}" && exit 1 )
```
It's book-data in the ruby script. Initially, I ran the ruby script to update books, so this became a problem.
```bash
all_books="cross-lib/book-data/AUTO_books.txt"
```

TylerZeroMaster commented on cross-lib/update_books.sh:
```diff
@@ -0,0 +1,24 @@
+
+# Refresh ABL
+abl="$(curl -Ss 'https://raw.githubusercontent.com/openstax/content-manager-approved-books/main/approved-book-list.json')"
+query=$(cat << 'EOF'
```
Both ways are **equally correct**, but I prefer simple single quotes to 'here documents' in this case.
```bash
# shellcheck disable=SC2016
query='
  .approved_books
  | .[]
  | .repository_name as $repo
  | .versions[0].commit_metadata.books
  | .[]
  | $repo + " " + .slug
'
```

kswanson33 replied:
(note: this file has been deleted)

TylerZeroMaster reviewed (commented):
I tried this out recently. Here are a few things I changed. Feel free pick and choose which changes you would like to keep, if any.

TylerZeroMaster commented on cross-lib/run_enki_on_all.sh:
```diff
@@ -0,0 +1,51 @@
+
+run_and_log_enki () {
+  echo "  running command $1 on $2 $3"
+  start_time=$(date +%s)
```
I forgot to mention this:
```bash
  data_dir="./data/$3-$1"
  [[ -d "$data_dir" ]] || mkdir -p "$data_dir"
  start_time=$(date +%s)
  ../enki --data-dir "$data_dir" --command "$1" --repo "openstax/$2" --book-slug "$3" --style default --ref main
```
Enk is up one directory and the data dirs were not being made.

kswanson33 replied:
ah - i have been testing this from enki root, i.e. running `./cross-lib/run_enki_on_all.sh --command all-git-web`. other paths also make this assumption.

kswanson33 commented:
current expected behavior:

input = 
```
tiny-book book-slug1
tiny-book broken
tiny-book book-slug1
```

output = 
<img width="742" alt="image" src="https://user-images.githubusercontent.com/26280712/227256154-ec6dac4a-a9b3-46ea-9161-a680fbc0f38a.png">

TylerZeroMaster commented on cross-lib/run_enki_on_all.sh:
```diff
@@ -0,0 +1,80 @@
+    --command) 
+      shift; arg_command=$1
+    ;;
+    *)
```
Some ideas for other options and implementations below:
```bash
    --echo)
      do_echo=true
    ;;
    --continue)
      do_continue=true
    ;;
```

TylerZeroMaster commented on cross-lib/run_enki_on_all.sh:
```diff
@@ -0,0 +1,80 @@
+  start_time=$(date +%s)
+  ./enki --data-dir ./data/$3-$1 --command $1 --repo openstax/$2 --book-slug $3 --style default --ref main &> $root/logs/$repo-$slug.txt
+  exit=$?
+  stop_time=$(date +%s)
```
```bash
  cmd="./enki --data-dir ./data/$3-$1 --command $1 --repo openstax/$2 --book-slug $3 --style default --ref main"
  log="$root/logs/$repo-$slug.txt"
  if [[ -f $log && $do_continue ]]; then return 0; fi
  if [[ $do_echo ]]; then
    $cmd 2>&1 | tee "$log"
  else
    $cmd &> "$log"
  fi
```

kswanson33 replied:
LOL good idea

TylerZeroMaster commented on cross-lib/update_books.rb:
```diff
@@ -0,0 +1,38 @@
+# Write ABL
+abl_hash['approved_books'].each do |book|
+  repo = book['repository_name']
+  book['versions'][0]['commit_metadata']['books'].each do |volume| 
```
One problem I ran into here is that the first entry in the version array tends to be the oldest entry (although there is nothing that guarantees this). The book slug in older versions of books on the ABL does not necessarily exist in the newest version of the repository (example: american-government vs american-government-3e).

Maybe you could store the `commit_sha` and use that when building books instead of `main`. Additionally, if you only want to build the latest version of each book, you could sort the `versions` by their `commit_metadata.committed_at` field.

kswanson33 replied:
thanks for the tip! i've gone with the latest version.

ultimately, i hope the slug will go away & books will be built by repo with this builder (pending [this](https://github.com/openstax/ce/issues/1971) 😅 ). but that can be a next pass
